block_conv.go

```diff
@@ -0,0 +1,266 @@
+package slack
+
+import (
+	"encoding/json"
+
+	"github.com/pkg/errors"
+)
+
+// UnmarshalJSON implements the Unmarshaller interface for Blocks, so that any JSON
+// unmarshalling is delegated and proper type determination can be made before unmarshal
+func (b *Blocks) UnmarshalJSON(data []byte) error {
+	var raw []json.RawMessage
+
+	if string(data) == "{\"blocks\":null}" {
+		return nil
+	}
+
+	err := json.Unmarshal(data, &raw)
+	if err != nil {
+		return err
+	}
+
+	var blocks Blocks
+	for _, r := range raw {
+		var obj map[string]interface{}
```

Comment:
define a type inside this method for this unmarshalling. generally avoid maps were possible.
```
type sumtype struct {
  blockType string `json:"type"`
}
s := sumtype{}
json.Unmarshal(r, &s)
```

Reply:
you could also define this type and embedded it in the block types. ./shrug depends how much it'll be used.

Reply:
changed from map to struct